API doc updates

yecril71pl opened the pull request:
# API doc updates
<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [X] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [X] You've included unit or integration tests for your change, where applicable.
- [X] You've included inline docs for your change, where applicable.
- [X] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Many unrelated changes manually merged into one PR for reasons I fail to understand.

## Description
1. Descriptions for [status codes](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.statuscodes) are available from [GetReasonPhrase](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.webutilities.reasonphrases.getreasonphrase).
2. Be specific about what a [Stream](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream) required by a [StreamOutputFormatter](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.formatters.streamoutputformatter) is.
3. A standard [StringOutputFormatter](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.formatters.stringoutputformatter) can only format a `string` into media type `text/plain` encoded with [UTF8](https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.utf8) or [Unicode](https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.unicode).
4. Provide stand-alone documentation for `ExecuteResultAsync` in [PageResult](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.razorpages.pageresult).
5. Add a link to [StatusCodes](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.statuscodes) to the description of [ContentResult.StatusCode](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.contentresult.statuscode#definition).
6. Add a link to [ResponseHeaders.ContentType](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.headers.responseheaders.contenttype) to the description of [ContentResult.ContentType](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.contentresult.contenttype#definition).
7. Mention that the magic directory of `DisplayTemplates` should be placed under [Pages](https://docs.microsoft.com/en-us/aspnet/core/tutorials/razor-pages/razor-pages-start#pages-folder-2).
8. Explain how to find the [template name](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.modelbinding.modelmetadata.templatehint) in the model metadata.
9. Explain how to specify a [template hint](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.modelbinding.modelmetadata.templatehint) in the model metadata.

Fixes #38833

yecril71pl committed:
Mention GetReasonPhrase in remarks

mkArtakMSFT commented:
Thanks for your PR, @yecril71pl.
Please combine your changes into a single PR and adjust the styling to match the rest of each document, as I've commented in the other PR.

yecril71pl commented:
> Thanks for your PR, @yecril71pl. Please combine your changes into a single PR and adjust the styling to match the rest of each document, as I've commented in the other PR.

Having or not having a space is not *styling*, it is *coding style*.  Please acknowledge that `<includedoc />` has a space in a place you do not like, which invalidates your consistent style hypothesis.

halter73 commented on src/Http/Http.Abstractions/src/StatusCodes.cs:
```diff
@@ -4,10 +4,13 @@
-///
-/// Status Codes listed at http://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
+/// A collection of constants for
+/// <see href="http://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml" >HTTP status codes</see >.
```
I'm hesitant to use `<see href=...` in a bunch of places since it breaks the summary in intellisense. See https://github.com/dotnet/aspnetcore/pull/38659#discussion_r760672551

yecril71pl replied:
I thought it is broken when the link is relative?

Tratcher replied:
I wasn't able to reproduce this issue, the href worked fine for me locally in VS 2022.

halter73 commented on src/Http/Http.Abstractions/src/StatusCodes.cs:
```diff
@@ -4,10 +4,13 @@
+/// </remarks>
 public static class StatusCodes
 {
     /// <summary>
```
Do you plan to include the description in each constant's `<summary>` like @Tratcher suggested in https://github.com/dotnet/aspnetcore/issues/38833#issuecomment-987033883? That seems like it could be helpful.

yecril71pl replied:
It would be more helpful is to refer to the documentation of [HttpStatusCode](https://docs.microsoft.com/en-us/dotnet/api/system.net.httpstatuscode) because 1DR > NIH.  But then, why not borrow the values defined in the system library?  I do not get that rejection at all.

Tratcher replied:
What's 1DR?

API cross references should be limited to APIs that will be directly used together. StatusCodes and HttpStatusCodes won't be used together. It's better for each of them to reference something authoritative like an RFC.

yecril71pl replied:
1DR is the one definition rule.  According to this rule, [StatusCodes](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.statuscodes), defined a dependent library, is secondary to [HttpStatusCode](https://docs.microsoft.com/en-us/dotnet/api/system.net.httpstatuscode), defined in the base library, and should import the values defined there rather than rolling its own (NIH).  It does not imply that [HttpStatusCode](https://docs.microsoft.com/en-us/dotnet/api/system.net.httpstatuscode) should be used alongside [StatusCodes](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.statuscodes); just the opposite, except in the definition of the latter (or where you need a true `enum` for some reason, which may also happen). And the documentation of the dependent library, again by 1DR, should reference the documentation of the base library, except for bogus values that are not defined there.

martincostello commented:
> Please acknowledge that `<includedoc />` has a space in a place you do not like, which invalidates your consistent style hypothesis.

Spaces are common within _self-closing_ tags like `<includedoc />`. They are not common in tags with inner text, like `<see href="foo">bar</see>`.

yecril71pl commented:
> Thanks for your PR, @yecril71pl. Please combine your changes into a single PR and adjust the styling to match the rest of each document, as I've commented in the other PR.

Could you please list the PRs you want me to combine?  I am confused by your request; it is an uncommon practice to combine unrelated changes.

yecril71pl commented:
> > Please acknowledge that `<includedoc />` has a space in a place you do not like, which invalidates your consistent style hypothesis.
> 
> Spaces are common within _self-closing_ tags like `<includedoc />`. They are not common in tags with inner text, like `<see href="foo">bar</see>`.

Does that also mean `<see cref="foo" />` is OK?  You seemed to be cross with it too.

mkArtakMSFT commented:
> > Thanks for your PR, @yecril71pl. Please combine your changes into a single PR and adjust the styling to match the rest of each document, as I've commented in the other PR.
> 
> Could you please list the PRs you want me to combine? I am confused by your request; it is an uncommon practice to combine unrelated changes.

[All the open PRs that you've sent](https://github.com/dotnet/aspnetcore/pulls/yecril71pl) are api doc updates and have similar comments. So would be great to combine into one to make sure the comments are also applied in unified way:
![image](https://user-images.githubusercontent.com/34246760/145072761-98afe815-8ebd-4706-ac6e-894dfd7abb46.png)

yecril71pl committed:
StreamOutputFormatter doc: define Stream

yecril71pl commented:
> [All the open PRs that you've sent](https://github.com/dotnet/aspnetcore/pulls/yecril71pl) are api doc updates and have similar comments. So would be great to combine into one to make sure the comments are also applied in unified way: ![image](https://user-images.githubusercontent.com/34246760/145072761-98afe815-8ebd-4706-ac6e-894dfd7abb46.png)

Greatness is in the eye of the beholder. You are making me do a lot of work just to make you feel good. But it seems I have no other choice.

yecril71pl committed:
StringOutputFormatter: explain how it works

yecril71pl committed:
PageResult doc: Describe ExecuteResultAsync

yecril71pl committed:
ContentResult doc: add links to properties

yecril71pl committed:
ModelMetadata doc: Use UIHint for TemplateHint

yecril71pl committed:
IHtmlHelper doc: DisplayTemplates are under Pages

mkArtakMSFT commented:
@yecril71pl that wasn't not my intention, it was just a way to avoid repetitive comments.
Thank you for combing the changes - it'll be easier to share feedback on similar things this way.

Tratcher commented on src/Mvc/Mvc.Core/src/ContentResult.cs:
```diff
@@ -19,12 +19,13 @@ public class ContentResult : ActionResult, IStatusCodeActionResult
     /// <summary>
-    /// Gets or sets the Content-Type header for the response.
+    /// Gets or sets the Content-Type header for the response,
+    /// to be used as <see cref="Microsoft.AspNetCore.Http.Headers.ResponseHeaders.ContentType" />.
```
How does this cross reference improve understanding? This property refers to a generic HTTP Content-Type header. ResponseHeaders.ContentType is just another property that refers to the same HTTP header, it's no more or less authoritative. They aren't even used together in the implementation.  An authoritative reference like https://www.rfc-editor.org/rfc/rfc4021.html#section-2.2.5 might be more helpful.

yecril71pl replied:
Prefer `cref` to `href` (because it shows up in Visual Studio) and prefer to stay on the documentation site (better UX).  I shall use `href` if you insist, please confirm!

Tratcher replied:
Referencing ResponseHeaders.ContentType from here isn't useful, they're barely related. I'm not aware of any local resource you could cref that would be useful, so either we don't add any link here, or we add an href to an external resource.

yecril71pl replied:
You can get to [MediaTypeHeaderValue](https://docs.microsoft.com/en-us/dotnet/api/microsoft.net.http.headers.mediatypeheadervalue) from [ContentType](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.headers.responseheaders.contenttype), which is what you want to use when you need to make a `Content-Type` header.  We could refer to that one directly if you prefer.
Your hyperlink goes to a standard document.  The purpose of a library is to isolate ordinary developers from having to read standard documents that are written for library developers.  Let&rsquo;s be merciful and not make the poor developers read that.

Tratcher replied:
These two properties wouldn't be used together in a developer workflow, they're both mainly for assignment and are for people working in different contexts. Please remove the link.

Tratcher commented on src/Mvc/Mvc.Core/src/Formatters/StringOutputFormatter.cs:
```diff
@@ -14,7 +14,10 @@ namespace Microsoft.AspNetCore.Mvc.Formatters;
 {
     /// <summary>
-    /// Initializes a new <see cref="StringOutputFormatter"/>.
+    /// A brand new <see cref="StringOutputFormatter"/>
+    /// supports only plain text to be encoded as
+    /// <see cref="P:System.Text.Encoding.UTF8" />
+    /// or <see cref="P:System.Text.Encoding.Unicode" />.
```
```suggestion
    /// Creates a new <see cref="StringOutputFormatter"/> that only supports plain text encoded as <see cref="P:System.Text.Encoding.UTF8" />
    /// or <see cref="P:System.Text.Encoding.Unicode" />.
```

Tratcher commented on src/Mvc/Mvc.Core/src/Formatters/StringOutputFormatter.cs:
```diff
@@ -23,7 +26,11 @@ public StringOutputFormatter()
 
-    /// <inheritdoc/>
+    /// <summary>
+    /// Verifies that the object to be formatted is a <see langword="string" />
+    /// and proceeds with the standard checks of
```
```suggestion
    /// Verifies that the object to be formatted is a <see langword="string" /> and proceeds with the standard checks of
```

Tratcher commented on src/Mvc/Mvc.RazorPages/src/PageResult.cs:
```diff
@@ -39,7 +39,18 @@ public class PageResult : ActionResult
+    /// </summary>
+    /// <param name="context">A <see cref="T:Microsoft.AspNetCore.Mvc.RazorPages.PageContext" />
+    /// associated with the current request for a Razor page.</param >
+    /// <returns >A <see cref="T:System.Threading.Tasks.Task" />
+    /// which will complete when page execution is completed.</returns >
+    /// <exception cref="T:System.ArgumentException"
+    /// >The parameter <paramref name="context" /> was not a
```
```suggestion
    /// <returns >A <see cref="T:System.Threading.Tasks.Task" /> which will complete when page execution is completed.</returns >
    /// <exception cref="T:System.ArgumentException">The parameter <paramref name="context" /> was not a
```

Tratcher commented on src/Mvc/Mvc.ViewFeatures/src/Rendering/IHtmlHelperOfT.cs:
```diff
@@ -70,7 +70,9 @@ public interface IHtmlHelper<TModel> : IHtmlHelper
     /// <para>
-    /// Custom templates are found under a <c>DisplayTemplates</c> folder. The folder name is case-sensitive on
+    /// Custom templates are found under a <c>DisplayTemplates</c> folder within the
+    /// <see href="https://docs.microsoft.com/en-us/aspnet/core/tutorials/razor-pages/razor-pages-start#pages-folder-2" >Pages</see > folder.
```
```suggestion
    /// <see href="https://docs.microsoft.com/aspnet/core/tutorials/razor-pages/razor-pages-start#pages-folder-2" >Pages</see > folder.
```

That said, we avoid putting direct concept doc links into the API docs because they may change. If we want to link to this resource we'd use an aka.ms link. @mkArtakMSFT is that something you can help with?

mkArtakMSFT replied:
Here you go: https://aka.ms/aspnet/7.0/razorpages-pages-folder

Note, that I've made this 7.0-specific, as the content may change in the future releases, and we may need to target different pages from the same API docs eventually.

Tratcher commented on src/Mvc/Mvc.ViewFeatures/src/Rendering/IHtmlHelperOfT.cs:
```diff
@@ -70,7 +70,9 @@ public interface IHtmlHelper<TModel> : IHtmlHelper
-    /// Custom templates are found under a <c>DisplayTemplates</c> folder. The folder name is case-sensitive on
+    /// Custom templates are found under a <c>DisplayTemplates</c> folder within the
+    /// <see href="https://docs.microsoft.com/en-us/aspnet/core/tutorials/razor-pages/razor-pages-start#pages-folder-2" >Pages</see > folder.
+    /// The folder name is case-sensitive on
     /// case-sensitive file systems.
```
```suggestion
    /// The folder name is case-sensitive on case-sensitive file systems.
```

yecril71pl committed:
Creates a new StringOutputFormatter

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

yecril71pl committed:
merge lines

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

yecril71pl committed:
merge lines

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

yecril71pl committed:
language-agnostic hyperlink

Note: I really prefer it to be en-us because the translations are hardly intelligible, so the first thing I do is to switch to the original, which is only one link away but the time and bandwidth is wasted.

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

yecril71pl committed:
merge lines

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

yecril71pl commented:
> @yecril71pl that wasn't not my intention, it was just a way to avoid repetitive comments. Thank you for combing the changes - it'll be easier to share feedback on similar things this way.

I combed through the changes and combined them :laughing:

yecril71pl committed:
handled using MediaTypeHeaderValue

Tratcher commented on src/Mvc/Mvc.ViewFeatures/src/Rendering/IHtmlHelperOfT.cs:
```diff
@@ -70,8 +70,9 @@ public interface IHtmlHelper<TModel> : IHtmlHelper
-    /// Custom templates are found under a <c>DisplayTemplates</c> folder. The folder name is case-sensitive on
-    /// case-sensitive file systems.
+    /// Custom templates are found under a <c>DisplayTemplates</c> folder within the
+    /// <see href="https://docs.microsoft.com/aspnet/core/tutorials/razor-pages/razor-pages-start#pages-folder-2" >Pages</see > folder.
```
```suggestion
    /// <see href="https://aka.ms/aspnet/7.0/razorpages-pages-folder">Pages</see> folder.
```

Tratcher committed:
Update src/Mvc/Mvc.ViewFeatures/src/Rendering/IHtmlHelperOfT.cs

Tratcher commented:
@yecril71pl Thanks for working through this with us.

Please do not send any more unsolicited PRs, nor is an appropriate venue for ideological debates. If there are things you find in the code base that need work please file issues first so we can discuss possible solutions in advance.

In the meantime, there are some existing issues that we could use your help on. See https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md#resources-to-help-you-get-started.

yecril71pl commented:
> In the meantime, there are some existing issues that we could use your help on. See https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md#resources-to-help-you-get-started.

How do I find issues regarding documentation?

Tratcher commented:
They'd be listed in the same Help Wanted queries, and also have the 'Docs' label.
